### PR TITLE
Application support in GraphQL v2

### PR DESCRIPTION
**Pull request by znarf**

### CODE REVIEW

File: server/graphql/v2/input/ApplicationReferenceInput.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ */
+export const fetchApplicationWithReference = async input => {
+  let application;
+  if (input.id) {
+    const id = idDecode(input.id, IDENTIFIER_TYPES.APPLICATION);
+    application = await models.Application.findByPk(id);
+  } else if (input.legacyId) {
+    application = await models.Application.findByPk(input.legacyId);
+  } else if (input.clientId) {
+    application = await models.Application.findOne('clientId', input.clientId);
```
Comment (znarf) [inline server/graphql/v2/input/ApplicationReferenceInput.js:38]:
```suggestion
    application = await models.Application.findOne({where: { 'clientId': input.clientId });
```

**Review by znarf — COMMENTED** [review]
Missing update.

File: server/graphql/v2/mutation/ApplicationMutations.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    },
+  },
+  async resolve(_, args, req) {
+    if (!req.remoteUser) {
+      throw new Unauthorized('You need to be authenticated to delete an application.');
+    }
+
+    const app = await fetchApplicationWithReference(args.application);
+    if (!app) {
+      throw new NotFound(`Application with id ${args.id} not found`);
```
Comment (znarf) [inline server/graphql/v2/mutation/ApplicationMutations.js:55]: Wrong error message if args != id

File: server/graphql/v2/query/ApplicationQuery.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+import { ApplicationReferenceInput, fetchApplicationWithReference } from '../input/ApplicationReferenceInput';
+import { Application } from '../object/Application';
+
+const ApplicationQuery = {
+  type: Application,
+  args: {
+    application: {
+      type: new GraphQLNonNull(ApplicationReferenceInput),
+      description: 'Identifiers to retrieve the Order',
```
Comment (Betree) [inline server/graphql/v2/query/ApplicationQuery.js:11]:
```suggestion
      description: 'Identifiers to retrieve the Application',
```

File: server/graphql/v2/input/ApplicationReferenceInput.js (hunk 1 above)
Comment (Betree) [inline server/graphql/v2/input/ApplicationReferenceInput.js:38]: Syntax?

File: server/graphql/v2/mutation/index.js [hunk 4]
```diff
@@ -1,5 +1,6 @@
 import accountMutations from './AccountMutations';
 import { addFundsMutation } from './AddFundsMutations';
+import applicationMutations from './ApplicationMutations';
```
Comment (Betree) [inline server/graphql/v2/mutation/index.js:3]: We're doing `export const` in there so we probably need

```suggestion
import * as applicationMutations from './ApplicationMutations';
```

Comment (Betree) [reply server/graphql/v2/mutation/index.js:3]: But we'll need to rename the exports if we do that

File: server/graphql/v2/input/ApplicationCreateInput.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+import URL from '../scalar/URL';
+
+export const ApplicationCreateInput = new GraphQLInputObjectType({
+  name: 'ApplicationCreateInput',
+  description: 'Input type for Application',
+  fields: () => ({
+    name: { type: GraphQLString },
+    description: { type: GraphQLString },
+    callbackUrl: { type: URL },
```
Comment (Betree) [inline server/graphql/v2/input/ApplicationCreateInput.js:11]: Should we add the type here?

Comment (znarf) [reply server/graphql/v2/input/ApplicationCreateInput.js:11]: What type?

File: server/graphql/v2/mutation/ApplicationMutations.js [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+export const updateApplicationMutation = {
+  type: Application,
+  args: {
+    application: {
+      type: new GraphQLNonNull(ApplicationUpdateInput),
+    },
+  },
+  async resolve(_, args, req) {
+    if (!req.remoteUser) {
+      throw new Unauthorized('You need to be authenticated to delete an application.');
```
Comment (Betree) [inline server/graphql/v2/mutation/ApplicationMutations.js:52]:
```suggestion
      throw new Unauthorized('You need to be authenticated to update an application.');
```

File: server/graphql/v2/mutation/index.js (hunk 4 above)
Comment (znarf) [reply server/graphql/v2/mutation/index.js:3]: I moved back to the approach we had in other files.

File: server/graphql/v2/input/ApplicationCreateInput.js (hunk 5 above)
Comment (Betree) [reply server/graphql/v2/input/ApplicationCreateInput.js:11]: Application type (oauth)

Comment (Betree) [reply server/graphql/v2/input/ApplicationCreateInput.js:11]: Even if it has only one value for now, to make it explicit that we're creating an oauth app, not an API key

**Pull request merged**